### DISCUSSION TITLE
Database integration

### DISCUSSION THREAD

**Discussion by divine** [clue/framework-x#70] [Q&A] [answered]
Hello,

Is there any chance that you could add at least one documentation for database integration (SQLite for instance?).

I've tried integrating it but simply failed, the query runs fine but the response doesn't run and is simply stuck (later killed due to 60 seconds limit of the process).

// Edit: not sure if it's Windows related.

Thanks!

**Comment by clue** [comment, marked as answer]
@divine Thanks for bringing this up, excellent question!

I'm glad to hear you're interested and already support Framework X as a sponsor, so here's an excerpt form the upcoming documentation:

```php
$factory = new Clue\React\SQLite\Factory();
$db = $factory->openLazy('count.db', null, ['idle' => 0.001]);

$app->get('/count', function (ServerRequestInterface $request) use ($db) {
    return $db->query('SELECT COUNT(*) AS count FROM hits')->then(function (Result $result) {
        return new Response(200, ['Content-Type' => 'text/plain'], $result->rows[0]['count'] . "\n");
    });
});
```

We're actually working on this very section in the documentation as we speak, but covering all the nasty details and possible options actually takes a fair amount of time, so the finished documentation might still take a while. I hope the above gist helps you get started!

Both the SQLite and MySQL client implementations have an idle time they use to keep a pending database connection alive in case you send another query within this period. If you're using the built-in web server, you can take advantage of this and have faster response times because the connection will be reused. However, if you're using this behind a traditional web server, it may actually keep the loop running and not end the pending request, thus slowing down the complete response. Given using this behind traditional web servers is becoming increasingly more common, I'm working on assigning better defaults for the upcoming versions of both database implementations 👍

**Reply by divine** [reply]
Hello @clue,

Thank you so much for your reply.

I had the same code that you gave as an example, let me try to clarify a little bit:

> here's an excerpt form the upcoming documentation

When I tried that it was giving me an error (Loop required), then I've found that documentation was for a new version that isn't released just yet: https://github.com/clue/reactphp-sqlite.

Using the master version fixed it.

Queries were running fine, just it wasn't completing the request.

I think you've fixed in the latest version but let me try that tomorrow.

> If you're using the built-in web server, you can take advantage of this and have faster response times because the connection will be reused.

The problem with the built-in web server is that I can't get the remote IP address of the visitor since $_SERVER['REMOTE_ADDR'] is empty as well as a user agent.

Do you have any suggestions for that?

Thanks!

**Reply by clue** [reply]
> > here's an excerpt form the upcoming documentation
> 
> When I tried that it was giving me an error (Loop required), then I've found that documentation was for a new version that isn't released just yet: https://github.com/clue/reactphp-sqlite.

Ouch, sorry for the confusion, my snippet indeed targets the current development version. You're not the first to stumble upon this (https://github.com/clue/reactphp-sqlite/issues/40), we'll make sure to get the version out as soon as possible to reduce friction here 👍

> > If you're using the built-in web server, you can take advantage of this and have faster response times because the connection will be reused.
> 
> The problem with the built-in web server is that I can't get the remote IP address of the visitor since $_SERVER['REMOTE_ADDR'] is empty as well as a user agent.

```php
$app->get('/user', function (Psr\Http\Message\ServerRequestInterface $request) {
    $params = $request->getServerParams();
    $ip = $params['REMOTE_ADDR'] ?? 'unknown';
    $agent = $request->getHeaderLine('User-Agent');

    return new React\Http\Message\Response(200, [], "Hello $agent from $ip\n");
});
```

See https://framework-x.clue.engineering/docs/api/request/#parameters and https://framework-x.clue.engineering/docs/api/request/#headers

Let me know if there's anything you feel we can do to improve the documentation here and I'm happy to take a look! 👍

**Reply by divine** [reply]
Hello @clue,
> Ouch, sorry for the confusion

No problem, I figured it on my own but thank your follow up :)


> See https://framework-x.clue.engineering/docs/api/request/#parameters and https://framework-x.clue.engineering/docs/api/request/#headers

This doesn't work behind a web server when run in production mode (ReactPHP/1) or put it simply all headers are lost when run in command:

`php app.php`

Simply results in:

```
{
	"REQUEST_TIME": 1632548796,
	"REQUEST_TIME_FLOAT": 1632548796.126677,
	"REMOTE_ADDR": "127.0.0.1",
	"REMOTE_PORT": 54442,
	"SERVER_ADDR": "127.0.0.1",
	"SERVER_PORT": 8080
}
```

As you can it doesn't contain important fields such as:

- HTTP_HOST
- **HTTP_USER_AGENT**
- HTTP_ACCEPT
- HTTP_ACCEPT_ENCODING
- HTTP_ACCEPT_LANGUAGE
- HTTP_COOKIE
- **HTTP_X_FORWARDED_FOR**
- HTTP_X_FORWARDED_PROTO

However, the PHP development server when run with the command  `php -S 0.0.0.0:8080 app.php`  contains all required fields and I don't think running that on the real web server will make sense at all.

Thanks!

**Reply by clue** [reply]
@divine I'm not sure I follow what problem you're seeing exactly, have you tried the above examples and the ones behind the above links?

* Instead of using `$_SERVER['HTTP_USER_AGENT']`, you should use `$request->getHeaderLine('User-Agent')`.
* Instead of using `$_SERVER['HTTP_X_FORWARDED_FOR']`, you should use `$request->getHeaderLine('X-Forwarded-For')`.
* Instead of using `$_SERVER['REMOTE_ADDR']`, you should use `$request->getServerParams()['REMOTE_ADDR']`.

All of these examples should work across all web servers, no matter whether you're running this in a traditional environment or using the built-in web server.

**Reply by divine** [reply]
@clue, I'm really sorry and I feel stupid for not trying that.

The difference is that the PHP development web server (when run with -S) prints those fields inside `$request->getServerParams()` as well but when running with _framework-x_ default web server there are no such fields inside `$request->getServerParams()` and I've never thought accessing those via `$request->getHeaderLine()`

Again, very much thank you for your help! I do really appreciate that!

Problem solved and I got what I needed!

Thanks!

**Reply by clue** [reply]
No worries, happy to hear we've got this sorted out now!

I can see where you're coming from. Please let me know (and/or file a PR?) if you feel there's anything we can improve in the documentation to help avoid this confusion.
